SolarEdge Local Component

drobtravels opened the pull request:
## Description:

Creates a new integration similar to the current SolarEdge component, except this communicates over a local API instead of cloud polling.  Note, this is **not** a replacement for the current SolarEdge component, as many SolarEdge Inverters do not  have this API.  Details are available in the documentation

**Related issue (if applicable):** fixes https://github.com/drobtravels/solaredge-local/issues/2

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#9490

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  platform: solaredge_local
  ip_address: 192.168.1.130
  monitored_conditions:
    - lifetime_energy
    - energy_this_year
    - energy_this_month
    - energy_today
    - current_power
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

drobtravels committed:
Basic local SolarEdge monitoring for energy / power

drobtravels committed:
Basic local SolarEdge monitoring for energy / power

drobtravels committed:
generated CODEOWNERS, requirements, excluded tests

drobtravels committed:
generated CODEOWNERS, requirements, excluded tests

drobtravels committed:
lint fixes, etc

drobtravels committed:
lint fixes, etc

homeassistant commented:
Hi @drobtravels, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#23996).

Once you do that we will be able to review and accept this pull request.

Thanks!

drobtravels committed:
Fix docstyle for init

Of course thats the file I forgot to run tests on

rohankapoorcom commented:
I don't think we want to have multiple components for this. It seems like the underlying library should accept a local mode or cloud mode flag and abstract away both implementations so that a single component in Home Assistant works.

drobtravels commented:
I'm happy to architect this however is recommended, however there are some big differences between the local API and cloud, which may make this component very messy.

To start with I exposed the same parameters to keep it simple, but the local API exposes a lot of details that the cloud version doesn't have such as optimizer level information, production meter information, and real time status.  Likewise cloud API has historical data not found in the local api.  Some users may want to use both together.

They also use different protocols.  The local API uses protocol buffers with no authentication, while the official cloud API supports JSON, XML, and CSV with API key authentication.

Due to these differences, they will likely require 2 independent python packages to communicate and different implementations.

Let me know how you recommend proceeding or if I can provide additional details.

drobtravels commented:
@rohankapoorcom Let me know if you have any further guidance based on my comments above.  If they were combined into a single component, I can think of two options for user configuration

1. Let the component decide which API to use for each sensor

```yaml
sensor:
  - platform: solaredge
    local_api:
      enabled: true
      ip_address: 192.168.1.130
    cloud_api:
      enabled: true
      api_key: API_KEY
      site_id: SITE_ID
    monitored_conditions:
        - lifetime_energy
        - energy_this_year
        - energy_this_month
        - energy_today
        - current_power
```

2.  Allow the user to specify monitored conditions for each API

```yaml
sensor:
  - platform: solaredge
    local_api:
      enabled: true
      ip_address: 192.168.1.130
      monitored_conditions:
        - lifetime_energy
        - energy_this_year
        - energy_this_month
        - energy_today
        - current_power
    cloud_api:
      enabled: true
      api_key: API_KEY
      site_id: SITE_ID
      monitored_conditions:
        - site_details
        - sensors
        - gateways
        - batteries
```

From an end user perspective, I think option 1 might be confusing as different monitored_conditions will be supported based on which APIs they configured.  Option 2 seems fine but doesn't seem any better then two separate components IMO.  From a code perspective, I think they should remain two different python packages.

pvizeli commented:
Based on upcoming ABR003 -> https://github.com/home-assistant/architecture/pull/244

Remove monitor condition. I think the local component is fine because it's a completely different protocol. I agree with @rohankapoorcom but it is how it is with smart homes...

drobtravels committed:
Load all sensors by default

They use the same API endpoint.  This changes was made per https://github.com/home-assistant/architecture/pull/244

drobtravels committed:
remve unneded date/time

drobtravels commented:
Removed monitored_conditions configuration options as requested per home-assistant/architecture#244

drobtravels committed:
ran hassfest again

pvizeli commented on homeassistant/components/solaredge_local/sensor.py:
```diff
@@ -0,0 +1,158 @@
+        self.api = api
+        self.data = {}
+
+    def update(self):
```
You need to add a trotthle here

pvizeli replied:
See on utils

drobtravels replied:
Will do, but out of curiosity, why is it necessary to throttle a local API?  I've been running it for several weeks without it.

pvizeli commented on .coveragerc:
```diff
@@ -548,7 +548,7 @@ omit =
     homeassistant/components/sochain/sensor.py
     homeassistant/components/socialblade/sensor.py
     homeassistant/components/solaredge/sensor.py
-    homeassistant/components/solax/sensor.py
```
why you remove solax?

drobtravels replied:
hmm.. Sorry about that.  I'm not used to the rebase workflow and must have messed something up.

drobtravels committed:
add throttle when updating

drobtravels committed:
readd solax, mistakenly removed

pvizeli committed:
Update sensor.py

Uskompuf commented:
Is this in release yet?

drobtravels commented:
This didn't make this release.  It should be part of 0.95 in ~2 weeks.
